File: src/familiar_connect/bus/protocols.py
"""Structural Protocols for sources, processors, bus itself.

Separate from concrete impls so future process-spanning ``EventBus``
drops in without touching processor code. See plan § Design.1,
§ Design.2.
"""

from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

    from familiar_connect.bus.envelope import Event


class BackpressurePolicy(Enum):
    """Per-topic behaviour when subscriber's queue full.

    - :attr:`BLOCK` — ``publish`` awaits until space frees. Use for
      data that must not be lost.
    - :attr:`DROP_OLDEST` — evict oldest queued, enqueue new. Use for
      hot streams where freshness beats completeness (audio).
    - :attr:`DROP_NEWEST` — drop incoming. Use when downstream stall
      must not affect other subscribers and late data worse than
      missing data.
    - :attr:`UNBOUNDED` — no backpressure; caller accepts memory risk.
      Use for low-volume channels where dropping is costly.
    """

    BLOCK = "block"
    DROP_OLDEST = "drop_oldest"
    DROP_NEWEST = "drop_newest"
    UNBOUNDED = "unbounded"


@runtime_checkable
class StreamSource(Protocol):
    """Produces events onto bus. Registered by runtime."""

    name: str

    async def run(self, bus: EventBus) -> None:
        """Run until bus drains; return cleanly on cancel."""
        ...


@runtime_checkable
class Processor(Protocol):
    """Subscribes to one or more topics; optionally re-publishes.

    ``topics`` consulted once at registration; dynamic subscription
    out of scope for v1.
    """

    name: str
    topics: tuple[str, ...]

    async def handle(self, event: Event, bus: EventBus) -> None:
        """Handle single event.

        Raised exceptions logged and swallowed by dispatcher — do not
        rely on them for flow control.
        """
        ...


class EventBus(Protocol):
    """Topic-addressed pub/sub surface."""

    async def start(self) -> None: ...

    async def shutdown(self) -> None: ...

    async def publish(self, event: Event) -> None: ...

    def subscribe(
        self,
        topics: tuple[str, ...],
        *,
        policy: BackpressurePolicy = BackpressurePolicy.BLOCK,
        maxsize: int = 0,
    ) -> AsyncIterator[Event]:
        """Async iterator of events for given topics.

        :param maxsize: ``0`` means default-for-policy.
        """
        ...

File: rust/familiar-connect/src/bus/protocols.rs
//! Bus seams: [`BackpressurePolicy`] + the [`EventBus`] / [`StreamSource`] /
//! [`Processor`] traits (subsystem 01; Python `bus/protocols.py`).
//!
//! Kept separate from the concrete [`InProcessEventBus`](crate::bus::InProcessEventBus)
//! so a future process-spanning `EventBus` drops in without touching processor
//! code. Python's runtime-checkable `Protocol`s become Rust traits (structural
//! conformance is enforced at compile time rather than by `isinstance`).

use std::sync::Arc;

use async_trait::async_trait;

use crate::bus::envelope::Event;
use crate::bus::in_process::Subscription;

/// Per-subscription behaviour when a subscriber's queue is full.
///
/// The wire string values (`"block"`, …) match the Python enum values.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub enum BackpressurePolicy {
    /// `publish` awaits until space frees — never loses data. The default.
    #[default]
    Block,
    /// Evict the oldest queued event, enqueue the new one (freshness beats
    /// completeness; e.g. audio).
    DropOldest,
    /// Drop the incoming event when full (a downstream stall must not affect
    /// other subscribers).
    DropNewest,
    /// No bound; the caller accepts the memory risk.
    Unbounded,
}

impl BackpressurePolicy {
    /// Every policy variant, in declaration order. Pins "exactly these four
    /// policies exist" (spec 01, `test_bus_protocols`).
    pub const ALL: [Self; 4] = [
        Self::Block,
        Self::DropOldest,
        Self::DropNewest,
        Self::Unbounded,
    ];

    /// The wire string value (matches the Python enum member values).
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Block => "block",
            Self::DropOldest => "drop_oldest",
            Self::DropNewest => "drop_newest",
            Self::Unbounded => "unbounded",
        }
    }
}

/// Topic-addressed pub/sub surface. **The** swappable seam — a future
/// cross-process bus drops in here (Python `EventBus` `Protocol`).
#[async_trait]
pub trait EventBus: Send + Sync {
    /// Transition `STARTING → RUNNING` (idempotent).
    async fn start(&self);
    /// Drain: close every subscription, then stop.
    async fn shutdown(&self);
    /// Fan an event out to every matching subscription, in registration order.
    async fn publish(&self, event: Event);
    /// Subscribe to `topics`. `maxsize == 0` means "default for policy".
    fn subscribe(
        &self,
        topics: &[&str],
        policy: BackpressurePolicy,
        maxsize: usize,
    ) -> Subscription;
}

/// Produces events onto the bus (Python runtime-checkable `StreamSource`).
#[async_trait]
pub trait StreamSource: Send + Sync {
    /// Stable identifier for logging/registration.
    fn name(&self) -> &str;
    /// Run until the bus drains; return cleanly on cancel.
    async fn run(&self, bus: Arc<dyn EventBus>);
}

/// Subscribes to one or more topics; optionally re-publishes (Python
/// runtime-checkable `Processor`).
///
/// Per DESIGN D6 (the resolved dispatcher error contract) `handle` returns a
/// `Result`, and the dispatch loop logs-and-continues on `Err` rather than
/// swallowing exceptions inside the processor — the behaviour the Python
/// docstring promised but the wiring never delivered.
#[async_trait]
pub trait Processor: Send + Sync {
    /// Stable identifier for logging/registration.
    fn name(&self) -> &str;
    /// Topics consulted once at registration (no dynamic subscription in v1).
    fn topics(&self) -> &[&str];
    /// Handle a single event. An `Err` is logged and swallowed by the dispatcher.
    async fn handle(&self, event: Arc<Event>, bus: &dyn EventBus) -> anyhow::Result<()>;
}
